Mathlib/ModelTheory/Semantics.lean

```diff
@@ -1023,6 +1042,301 @@ theorem infinite [Mi : Infinite M] (h : M ≅[L] N) : Infinite N :=
+      convert ih i b hb
+      funext sum
+      cases sum
+      · rfl
+      · simp
+        rw [Equiv.leftInverse_symm finSigmaFinEquiv]
+
+/-- `Term.subst_definitions_extraVals` bundled with their defining property
+  `Term.subst_definitions_extraVals_spec`. -/
+def subst_definitions_extraVals_X (t : L.Term α)
+  {Fs : ∀ {n} (_ : L.Functions n), L'.Formula (Fin n ⊕ Unit)}
+  (hFs : ∀ {n} (g : L.Functions n),
+    (Function.tupleGraph fun v ↦ g.term.realize v) = ((Fs g).Realize : Set (_ → M)))
+    (v : _)
+  : { xs : Fin (t.subst_definitions Fs).1 → M //
+    ∀ s ∈ (t.subst_definitions Fs).2.2, s.Realize (Sum.elim v xs)} :=
+  ⟨t.subst_definitions_extraVals Fs v, t.subst_definitions_extraVals_spec hFs v⟩
+
+end Term
+
+namespace BoundedFormula
+
+/-- `BoundedFormula.subst_definitions` agrees with the original formula once realized. -/
+theorem subst_definitions_eq {k : ℕ} (f : L.BoundedFormula α k)
+  {Fs : ∀ {n} (_ : L.Functions n), L'.Formula (Fin n ⊕ Unit)}
+  (hFs : ∀ {n} (g : L.Functions n),
+    (Function.tupleGraph fun v ↦ g.term.realize v) = ((Fs g).Realize : Set (_ → M)))
+  {Rs : ∀ {n} (_ : L.Relations n), L'.Formula (Fin n)}
+  (hRs : ∀ {n} (g : L.Relations n), RelMap (M := M) g = (Rs g).Realize)
+  : ∀ vL vR, (f.subst_definitions Fs Rs).Realize (M := M) vL vR = f.Realize vL vR := by
+    induction f
+    next =>
+      simp [Realize.eq_1, subst_definitions.eq_1]
+    next n t₁ t₂ =>
+      simp only [subst_definitions, finSumFinEquiv_apply_left, finSumFinEquiv_apply_right,
+        Nat.add_zero, realize_relabel, castAdd_zero, cast_refl, Function.comp_id, Realize,
+        eq_iff_iff]
+      intro vL vR
+      rw [show (vR ∘ @Fin.natAdd 0 n) = default from Unique.eq_default _]
+      rw [← Formula.Realize, realize_alls]
+      simp_rw [realize_foldr_imp]
+      simp_rw [BoundedFormula.Realize.eq_2]
+      constructor
+      · intro h
+        have ⟨xs₁, hxs₁⟩ := t₁.subst_definitions_extraVals_X hFs (Sum.elim vL vR)
+        have ⟨xs₂, hxs₂⟩ := t₂.subst_definitions_extraVals_X hFs (Sum.elim vL vR)
+        have ht₁ := t₁.subst_definitions_eq hFs (Sum.elim vL vR) hxs₁
+        have ht₂ := t₂.subst_definitions_eq hFs (Sum.elim vL vR) hxs₂
+        rw [← ht₁, ← ht₂]
+        clear ht₁ ht₂
+        replace h := h (Fin.appendEquiv _ _ ⟨xs₁,xs₂⟩) (by
+          intro i hi
+          simp only [List.mem_append, List.mem_map] at hi
+          rcases hi with (⟨w,hiw,rfl⟩|⟨w,hiw,rfl⟩) <;> {
+            try have := hxs₁ w hiw
+            try have := hxs₂ w hiw
+            rw [Formula.Realize] at this
+            rw [realize_relabel]
+            convert this
+            funext sum
+            cases sum <;> simp
+          }
+        )
+        convert h using 1 <;> {
+          rw [Term.realize_relabel]
+          congr
+          funext sum
+          cases sum <;> simp
+        }
+      · intro h xs is
+        have ht₁ := t₁.subst_definitions_eq hFs _ (by
+          simp at is
```

Comment:
This proof looks weird.

Reply:
Happy to change it if you tell me an equivalent/better way to write it. The actual statement of the have is quite long and ugly. I tried `:= t₁.subst_definitions_eq hFs _ ?_` so that I could fill in the missing proof as another goal, but the it can't infer the value of the underscore and it gives errors.